src/main/java/org/aksw/iguana/cc/query/handler/QueryHandler.java

```diff
@@ -176,37 +175,48 @@ public QueryHandler(Config config) throws IOException {
         this.config = config;
         var querySource = createQuerySource(Path.of(config.path));
 
-        if (config.pattern() != null) {
-            final var originalPath = querySource.getPath();
-            Path instancePath = Files.isDirectory(originalPath) ?
-                    originalPath.resolveSibling(originalPath.getFileName() + "_instances.txt") :
-                    originalPath.resolveSibling(originalPath.getFileName().toString().split("\\.")[0] + "_instances.txt");
-            if (Files.exists(instancePath)) {
-                LOGGER.info("Already existing query pattern instances have been found and will be reused. Delete the following file to regenerate them: {}", instancePath.toAbsolutePath());
-                querySource = createQuerySource(instancePath);
-            } else {
-                final List<String> instances = instantiatePatternQueries(querySource, config.pattern);
-                if (config.pattern.save) {
-                    Files.createFile(instancePath);
-                    try (var writer = Files.newBufferedWriter(instancePath)) {
-                        for (String instance : instances) {
-                            writer.write(instance);
-                            writer.newLine();
-                        }
+        // initialize queryList based on the given configuration
+        if (config.template() != null) {
+            queryList = initializeTemplateQueryHandler(querySource);
+        } else {
+            queryList = (config.caching()) ?
+                    new FileCachingQueryList(querySource) :
+                    new FileReadingQueryList(querySource);
+        }
+        this.hashCode = queryList.hashCode();
+    }
+
+    private QueryList initializeTemplateQueryHandler(QuerySource templateSource) throws IOException {
+        QuerySource querySource = templateSource;
+        final var originalPath = templateSource.getPath();
+        final Path instancePath = Files.isDirectory(originalPath) ?
+                originalPath.resolveSibling(originalPath.getFileName() + "_instances.txt") : // if the source of the query templates is a folder, the instances will be saved in a file with the same name as the folder
+                originalPath.resolveSibling(originalPath.getFileName().toString().split("\\.")[0] + "_instances.txt"); // if the source of the query templates is a file, the instances will be saved in a file with the same name as the file
+        if (Files.exists(instancePath)) {
+            LOGGER.info("Already existing query template instances have been found and will be reused. Delete the following file to regenerate them: {}", instancePath.toAbsolutePath());
+            querySource = createQuerySource(instancePath); // if the instances already exist, use them
```

Comment:
The path should contain not only the file-name but also contain the parameters:
endpoint and limit. For endpoint, use  a hash of the endpoint string. 

Can be moved to a separate low priority issue and resolved later.